File: PredictionAgent-Demo/src/nodes/data_fetch_node.py
"""
数据获取节点
负责从数据库获取产品相关的销售数据和预测数据
"""

from typing import Dict, Any
from datetime import date, timedelta
from .base_node import BaseNode
from ..database.repository import SalesRepository
from ..database.models import ProductAnalysisData


class DataFetchNode(BaseNode):
    """数据获取节点"""

    def __init__(self, llm_client, repository: SalesRepository):
        super().__init__(llm_client, "DataFetch")
        self.repository = repository

    def run(self, input_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        执行数据获取

        Args:
            input_data: {
                "product_code": str,
                "product_name": str,
                "history_days": int (可选，默认90天),
                "future_days": int (可选，默认30天)
            }

        Returns:
            {
                "fetched": bool,
                "historical_data": List[Dict],
                "model_predictions": List[Dict],
                "future_predictions": List[Dict],
                "statistics": Dict,
                "error_message": str
            }
        """
        product_code = input_data.get("product_code", "")

        if not product_code:
            return {
                "fetched": False,
                "historical_data": [],
                "model_predictions": [],
                "future_predictions": [],
                "statistics": {},
                "error_message": "产品代码为空"
            }

        history_days = input_data.get("history_days", 90)
        future_days = input_data.get("future_days", 30)

        self.log_info(f"正在获取产品 {product_code} 的数据...")

        try:
            # 获取完整的分析数据
            analysis_data = self.repository.get_product_analysis_data(
                product_code=product_code,
                history_days=history_days,
                future_days=future_days
            )

            # 转换为可序列化的格式
            historical_data = self._convert_historical_data(analysis_data)
            model_predictions = self._convert_predictions(analysis_data.model_predictions)
            future_predictions = self._convert_predictions(analysis_data.future_predictions)

            # 获取统计信息
            statistics = self.repository.get_product_statistics(product_code, history_days)

            result = {
                "fetched": True,
                "historical_data": historical_data,
                "model_predictions": model_predictions,
                "future_predictions": future_predictions,
                "statistics": statistics,
                "error_message": ""
            }

            self.log_info(f"数据获取成功: 历史数据 {len(historical_data)} 条, "
                         f"模型预测 {len(model_predictions)} 条, "
                         f"未来预测 {len(future_predictions)} 条")

            return result

        except Exception as e:
            self.log_error(f"数据获取失败: {str(e)}")
            return {
                "fetched": False,
                "historical_data": [],
                "model_predictions": [],
                "future_predictions": [],
                "statistics": {},
                "error_message": str(e)
            }

    def _convert_historical_data(self, analysis_data: ProductAnalysisData) -> list:
        """转换历史数据"""
        result = []
        sorted_sales = sorted(analysis_data.sales_history, key=lambda x: x.sale_date)

        # 构建预测值映射
        prediction_map = {}
        for p in analysis_data.model_predictions:
            date_str = p.prediction_date.isoformat() if p.prediction_date else ""
            prediction_map[date_str] = p.predicted_value

        for sale in sorted_sales:
            date_str = sale.sale_date.isoformat() if sale.sale_date else ""
            result.append({
                "date": date_str,
                "actual_value": sale.quantity,
                "predicted_value": prediction_map.get(date_str, None)
            })

        return result

    def _convert_predictions(self, predictions: list) -> list:
        """转换预测数据"""
        return [
            {
                "date": p.prediction_date.isoformat() if p.prediction_date else "",
                "predicted_value": p.predicted_value,
                "confidence": p.confidence
            }
            for p in sorted(predictions, key=lambda x: x.prediction_date)
        ]

    def fetch_mock_data(self, product_code: str, product_name: str,
                        history_days: int = 90, future_days: int = 30) -> Dict[str, Any]:
        """
        生成模拟数据（用于测试或没有真实数据时）

        Args:
            product_code: 产品代码
            product_name: 产品名称
            history_days: 历史天数
            future_days: 未来天数

        Returns:
            模拟数据
        """
        import random
        from datetime import datetime, timedelta

        today = date.today()
        historical_data = []
        future_predictions = []

        # 生成历史数据（带趋势和季节性）
        base_value = random.randint(80, 200)
        for i in range(history_days):
            d = today - timedelta(days=history_days - i - 1)
            # 模拟趋势
            trend = i * 0.3
            # 模拟周季节性
            weekday_factor = 1.2 if d.weekday() < 5 else 0.8
            # 模拟随机波动
            noise = random.uniform(0.9, 1.1)
            actual = int((base_value + trend) * weekday_factor * noise)

            # 模拟模型预测（略有偏差）
            predicted = actual * random.uniform(0.95, 1.05)

            historical_data.append({
                "date": d.isoformat(),
                "actual_value": actual,
                "predicted_value": round(predicted, 2)
            })

        # 生成未来预测
        for i in range(1, future_days + 1):
            d = today + timedelta(days=i)
            weekday_factor = 1.2 if d.weekday() < 5 else 0.8
            predicted = int((base_value + trend + i * 0.3) * weekday_factor)

            future_predictions.append({
                "date": d.isoformat(),
                "predicted_value": predicted,
                "confidence": round(0.95 - i * 0.01, 2)  # 越远期置信度越低
            })

        # 计算统计信息
        actual_values = [d["actual_value"] for d in historical_data]
        avg_daily = sum(actual_values) / len(actual_values)

        mid = len(actual_values) // 2
        first_half_avg = sum(actual_values[:mid]) / mid
        second_half_avg = sum(actual_values[mid:]) / (len(actual_values) - mid)
        trend_change = ((second_half_avg - first_half_avg) / first_half_avg) * 100 if first_half_avg > 0 else 0

        return {
            "fetched": True,
            "historical_data": historical_data,
            "model_predictions": [],  # 模拟数据中没有单独的模型预测
            "future_predictions": future_predictions,
            "statistics": {
                "product_code": product_code,
                "period_days": history_days,
                "avg_daily_sales": round(avg_daily, 2),
                "trend_change_percent": round(trend_change, 2),
                "trend_direction": "up" if trend_change > 5 else ("down" if trend_change < -5 else "stable")
            },
            "error_message": ""
        }

File: PredictionAgent-Go/nodes/data_fetch_node.go
package nodes

import (
	"context"
	"fmt"
	"math/rand"
	"time"

	"prediction-agent/llm"
)

// DataFetchNode 数据获取节点
type DataFetchNode struct {
	*BaseNode
	LLMClient llm.LLMProvider
}

// DataFetchResult 数据获取结果
type DataFetchResult struct {
	Fetched          bool              `json:"fetched"`
	HistoricalData   []map[string]any `json:"historical_data"`
	ModelPredictions []map[string]any `json:"model_predictions"`
	FuturePredictions []map[string]any `json:"future_predictions"`
	Statistics       map[string]any   `json:"statistics"`
	ErrorMessage    string            `json:"error_message"`
}

// NewDataFetchNode 创建数据获取节点
func NewDataFetchNode(llmClient llm.LLMProvider) *DataFetchNode {
	return &DataFetchNode{
		BaseNode:   NewBaseNode(llmClient, "DataFetch"),
		LLMClient: llmClient,
	}
}

// Run 执行数据获取
func (n *DataFetchNode) Run(ctx context.Context, inputData map[string]any) DataFetchResult {
	productCode, _ := inputData["product_code"].(string)
	productName, _ := inputData["product_name"].(string)
	historyDays := 90
	futureDays := 30

	if h, ok := inputData["history_days"].(int); ok {
		historyDays = h
	}
	if f, ok := inputData["future_days"].(int); ok {
		futureDays = f
	}

	if productCode == "" {
		return DataFetchResult{
			Fetched:         false,
			HistoricalData:  []map[string]any{},
			ErrorMessage:    "产品代码为空",
		}
	}

	n.LogInfo(fmt.Sprintf("正在获取产品 %s 的数据...", productCode))

	result := n.FetchMockData(productCode, productName, historyDays, futureDays)
	n.LogInfo(fmt.Sprintf("数据获取成功: 历史数据 %d 条, 未来预测 %d 条", 
		len(result.HistoricalData), len(result.FuturePredictions)))

	return result
}

// FetchMockData 生成模拟数据
func (n *DataFetchNode) FetchMockData(productCode, productName string, historyDays, futureDays int) DataFetchResult {
	rand.Seed(time.Now().UnixNano())
	
	historicalData := make([]map[string]any, 0, historyDays)
	futurePredictions := make([]map[string]any, 0, futureDays)
	
	baseValue := float64(rand.Intn(120) + 80)
	trend := 0.3
	
	now := time.Now()
	
	// 生成历史数据
	for i := 0; i < historyDays; i++ {
		d := now.AddDate(0, 0, -historyDays+i)
		
		// 模拟周季节性
		weekdayFactor := 1.2
		if d.Weekday() == 0 || d.Weekday() == 6 {
			weekdayFactor = 0.8
		}
		
		// 模拟随机波动
		noise := 0.9 + rand.Float64()*0.2
		actual := int((baseValue + float64(i)*trend) * weekdayFactor * noise)
		
		// 模拟模型预测
		predicted := float64(actual) * (0.95 + rand.Float64()*0.1)
		
		historicalData = append(historicalData, map[string]any{
			"date":           d.Format("2006-01-02"),
			"actual_value":   actual,
			"predicted_value": predicted,
		})
	}
	
	// 生成未来预测
	for i := 1; i <= futureDays; i++ {
		d := now.AddDate(0, 0, i)
		
		weekdayFactor := 1.2
		if d.Weekday() == 0 || d.Weekday() == 6 {
			weekdayFactor = 0.8
		}
		
		predicted := int((baseValue + float64(historyDays+i)*trend) * weekdayFactor)
		confidence := 0.95 - float64(i)*0.01
		if confidence < 0.7 {
			confidence = 0.7
		}
		
		futurePredictions = append(futurePredictions, map[string]any{
			"date":            d.Format("2006-01-02"),
			"predicted_value": predicted,
			"confidence":      confidence,
		})
	}
	
	// 计算统计信息
	actualValues := make([]float64, 0)
	for _, d := range historicalData {
		if v, ok := d["actual_value"].(int); ok {
			actualValues = append(actualValues, float64(v))
		}
	}
	
	avgDaily := 0.0
	if len(actualValues) > 0 {
		sum := 0.0
		for _, v := range actualValues {
			sum += v
		}
		avgDaily = sum / float64(len(actualValues))
	}
	
	mid := len(actualValues) / 2
	firstHalfAvg := sum(actualValues[:mid])
	secondHalfAvg := sum(actualValues[mid:])
	
	trendChange := 0.0
	if firstHalfAvg > 0 {
		trendChange = ((secondHalfAvg - firstHalfAvg) / firstHalfAvg) * 100
	}
	
	trendDirection := "stable"
	if trendChange > 5 {
		trendDirection = "up"
	} else if trendChange < -5 {
		trendDirection = "down"
	}
	
	statistics := map[string]any{
		"product_code":          productCode,
		"period_days":           historyDays,
		"avg_daily_sales":      avgDaily,
		"trend_change_percent":  trendChange,
		"trend_direction":       trendDirection,
	}
	
	return DataFetchResult{
		Fetched:           true,
		HistoricalData:   historicalData,
		ModelPredictions: []map[string]any{},
		FuturePredictions: futurePredictions,
		Statistics:       statistics,
		ErrorMessage:     "",
	}
}

func sum(values []float64) float64 {
	s := 0.0
	for _, v := range values {
		s += v
	}
	if len(values) > 0 {
		return s / float64(len(values))
	}
	return 0
}
